---
title: "Figure 5E and S10 - distributions of indel size"
author: "Vitalii Kleshchevnikov"
date: "06/07/2018"
output: 
  html_document: 
    toc: yes
---

```{r setup}
start_time = Sys.time()
source("./r_utils/installMissingPackages.R")
packages = c("R.utils", "data.table", "ggplot2", "GGally", "cowplot", "grid", "devtools", "svglite")
installMissingPackages(packages)

suppressPackageStartupMessages({
  library(R.utils)
  library(data.table)
  library(ggplot2)
  library(cowplot)
  library(grid)
})
source("./r_utils/myggpairs.R")
source("./r_utils/loadFromFARM.R")
source("./r_utils/readIndelSummaries.R") 
source("./r_utils/dominantIndel.R")
source("./r_utils/findIndelLevels.R")
source("./r_utils/Other.R")
source("./r_utils/addIndelRepl.R") 
source("./r_utils/selectCellLines.R")
source("./r_utils/predictedK562Files.R")
source("./r_utils/renameSampleGroup.R")
plot_count = 1
my.ggsave = function(plot_count, name = NULL, width = NA, height = NA,
                     units = c("in", "cm", "mm"), device = "svg"){
  if(is.null(name)) 
    name = paste0("../results/plots/", plot_count, ".", device) else
      name = paste0("../results/plots/", name, ".", device)
    dir = gsub(basename(name),"", name)
    if(!dir.exists(dir)) dir.create(dir, recursive = T)
    ggsave(name, device = device, width = width, height = height, units = units)
    plot_count + 1
} # usage plot_count = my.ggsave(plot_count)

## Find data
data.base.dir = "../data/summary_data/"
data.set.dir = "more_indel_summaries/"
data.dir = paste0(data.base.dir, data.set.dir) 
data.files = paste0(data.dir, list.files(data.dir))
```

## Load and process data

```{r load_all_cell_lines}
####/ set some options
N_reads_filter = 20
inv = F
# remove remove_2A_TREX2?
remove_2A_TREX2 = F

# define order of variables and colors for plots
source("./r_utils/order_var_script.R")

# choose thresholds for dominant indel
dominant_indel_freq = 0.2 # individual indels
dominant_indel_class_freq = 0.4 # all indels from one category
# normalise by the number of guides in each category in control K562?
normalise_guides = F
####\

# select relevant files:
new.DPI7.data.files = selectCellLinesDPI7(data.files)

real_guides_20 = fread("../data/summary_data/real_guides_in_dpi7_samples_above20mutreads.txt", stringsAsFactors = F, header = F)
data = readIndelSummaries(new.DPI7.data.files, guide_list = real_guides_20$V1, inv = inv)

# use N reads filter
data = inAllSamples(data[Total_reads >= N_reads_filter],.(Oligo_Id))

# how many guides per sample?
guidesPerSample(data)
Sys.time() - start_time
```

## Set options and select real guides represented in all samples above a certain threshold 

```{r data_processing}
#renames samples for publication
data = renameSampleGroup(data)
# calculate means of normalised and cumulative reads across replicates
data = addIndelRepl(data)

# sanity check - uncomment when reproducing using different data
#unique(data[Oligo_Id == Oligo_Id[1] & 
#              Most_Common_Indel == Most_Common_Indel[1] &
#              sample_group == "K562"]) # should be one line
#data[, sum(sum_MCI_Reads) > sum_Total_reads, by = .(Oligo_Id, sample_group)][,mean(V1)] # should be 0
```

## Calculate proportions of indels of each size

```{r prop_indels_by_size, eval=TRUE, fig.height=4.5, fig.width=10}
# filter data by indel class
data.indels.m = unique(data[indel_class %in% c(paste0("I",1:10), paste0("D",1:30)),
                             .(Oligo_Id, Most_Common_Indel, mean_Norm_MCI_Reads, Norm_MCI_Reads, few_indel_classes, sample_group, indel_class, Microhomology_Size)])

# set order of indel classes
indel_levels = findIndelLevels(data.indels.m, gap = T)
data.indels.m[, indel_class := factor(indel_class,
                                      levels = indel_levels)]
# sum up reads for indels of the same class within samples
data.indels.m[, reads_per_indel_class := sum(mean_Norm_MCI_Reads), 
              by = c("Oligo_Id", "sample_group", "indel_class")]

# for each indel class in control K562 calculate sum/mean/median (across guides) the frequency of outcomes of each class (reads_per_indel_class) in other sample groups
data.indels3 = unique(data.indels.m[,.(Oligo_Id, indel_class, sample_group, reads_per_indel_class)])

data.indels3[, indel_class_mean_general := as.numeric(mean(reads_per_indel_class)),
            by = .(indel_class, sample_group)]

legend_lab4 = paste0("\n(",guidesPerSampleGroup(data.indels3)[, unique(N_guides)]," total)")


# set order of samples
data.indels3[, sample_group := factor(sample_group,
                                             levels = sample_group_levels)]
# add rows to add a space between insertions and deletions in figures
data.indels3 = rbind(data.indels3, unique(data.table(indel_class = "",
                              indel_class_mean_general = NA,
                              sample_group = data.indels3$sample_group,
                              Oligo_Id = data.indels3$Oligo_Id,
                              reads_per_indel_class = NA)))
```

## Figure 5E. Cas9-TREX2 fusion protein favours larger deletions compared to K562. 

Deletions of increasing size (x-axis) become more frequent (y-axis) in K562 Cas9-TREX2 cells (blue) compared to standard K562 Cas9 (orange). gRNAs as in A.

```{r histog_detailed_outcomes_trex2, eval=TRUE, fig.height=3, fig.width=7.4}
TREX = data.indels3[sample_group %in% c("K562", "TREX2")]
# sparsify axis labels
TREX[, indel_class_lab := gsub("D|I", "", indel_class)]
TREX[as.numeric(indel_class_lab) > 10 & as.numeric(indel_class_lab) %% 2 == 1, indel_class_lab := ""]
TREX.hist = unique(TREX[,.(indel_class, indel_class_mean_general, indel_class_lab, sample_group)])

# overlapping histograms Figure 5E
TREX.bar.plot3 = ggplot(aes(x = indel_class, y = indel_class_mean_general,
                           color = sample_group, group = sample_group), 
                       data = unique(TREX.hist[,.(indel_class, indel_class_mean_general,
                                             sample_group)])) +
  theme_bw() +
  geom_line(size = 1.5) + geom_point(size = 2.5) +
  ylab("Mean frequency of outcome") +
  theme(plot.title = element_text(size=14,face="bold"),
        axis.title = element_text(size=14,face="bold"),
        axis.title.y = element_text(size=12),
        axis.title.x = element_text(hjust = 1.2,size=0),
        axis.text = element_text(size = 10),
        axis.text.x = element_text(angle = 0, hjust = 0.5, vjust = 0.5),
        legend.position = c(0.9,0.87),
        strip.text.y = element_text(angle = 0),
        panel.grid.minor = element_line(size = 0),
        panel.grid.major.x = element_line(size = 0),
        plot.background = element_rect(fill="transparent", colour="transparent"),
        legend.background = element_rect(fill="transparent", colour="transparent")) +
  scale_x_discrete(labels = TREX$indel_class_lab,
                   breaks = TREX$indel_class) +
  labs(fill = "", color = "")

{
  grid.newpage()
  ggdraw() +
    draw_plot(TREX.bar.plot3 +
                scale_color_manual(values=trex_colors)+
                scale_fill_manual(values=trex_colors),
              x = 0, y = 0.1, width = 1, height = 0.9) +
    draw_plot(grid.lines(x = c(0,5.45), y = c(0.01, 0.01),
                         gp=gpar(size=2)),
              x = .09, y = 0.1, width = .12, height = 1) +
    draw_plot(textGrob("Deletions", gp=gpar(fontsize=12)),
              x = .35, y = -0.08, width = .15, height = .3) +
    draw_plot(grid.lines(x = c(0,1.83), y = c(0.01, 0.01),
                         gp=gpar(size=2)),
              x = .77, y = 0.1, width = .12, height = 1) +
    draw_plot(textGrob("Insertions", gp=gpar(fontsize=12)),
              x = .8, y = -0.08, width = .15, height = .3)
  }
plot_count = my.ggsave(plot_count, name = "Fig_5E", width = 7.4, height = 3, units = "in")
```


## Figure S18. Cas9-2A-TREX2 has a different influence on repair outcomes compared to Cas9-TREX2

The mean frequency (y-axis) of deletion or insertion size (x-axis) across genomic sequence targets for three alternative Cas9 effector constructs (colors).

```{r histog_detailed_outcomes_S10, eval=TRUE, fig.height=3, fig.width=7.4}
TREX = data.indels3[sample_group %in% c("K562", "TREX2", "2A_TREX2")]
# sparsify axis labels
TREX[, indel_class_lab := gsub("D|I", "", indel_class)]
TREX[as.numeric(indel_class_lab) > 10 & as.numeric(indel_class_lab) %% 2 == 1, indel_class_lab := ""]
TREX.hist = unique(TREX[,.(indel_class, indel_class_mean_general, indel_class_lab, sample_group)])

# overlapping histograms Figure 5E
TREX.bar.plot3 = ggplot(aes(x = indel_class, y = indel_class_mean_general,
                           color = sample_group, group = sample_group), 
                       data = unique(TREX.hist[,.(indel_class, indel_class_mean_general,
                                             sample_group)])) +
  theme_bw() +
  geom_line(size = 1.5) + geom_point(size = 2.5) +
  ylab("Mean frequency of outcome") +
  theme(plot.title = element_text(size=14,face="bold"),
        axis.title = element_text(size=14,face="bold"),
        axis.title.y = element_text(size=12),
        axis.title.x = element_text(hjust = 1.2,size=0),
        axis.text = element_text(size = 10),
        axis.text.x = element_text(angle = 0, hjust = 0.5, vjust = 0.5),
        legend.position = c(0.9,0.87),
        strip.text.y = element_text(angle = 0),
        panel.grid.minor = element_line(size = 0),
        panel.grid.major.x = element_line(size = 0),
        plot.background = element_rect(fill="transparent", colour="transparent"),
        legend.background = element_rect(fill="transparent", colour="transparent")) +
  scale_x_discrete(labels = TREX$indel_class_lab,
                   breaks = TREX$indel_class) +
  labs(fill = "", color = "")

{
  grid.newpage()
  ggdraw() +
    draw_plot(TREX.bar.plot3 +
                scale_color_manual(values=trex_colors)+
                scale_fill_manual(values=trex_colors),
              x = 0, y = 0.1, width = 1, height = 0.9) +
    draw_plot(grid.lines(x = c(0,5.45), y = c(0.01, 0.01),
                         gp=gpar(size=2)),
              x = .09, y = 0.1, width = .12, height = 1) +
    draw_plot(textGrob("Deletions", gp=gpar(fontsize=12)),
              x = .35, y = -0.08, width = .15, height = .3) +
    draw_plot(grid.lines(x = c(0,1.83), y = c(0.01, 0.01),
                         gp=gpar(size=2)),
              x = .77, y = 0.1, width = .12, height = 1) +
    draw_plot(textGrob("Insertions", gp=gpar(fontsize=12)),
              x = .8, y = -0.08, width = .15, height = .3)
  }
plot_count = my.ggsave(plot_count, name = "Fig_S18", width = 7.4, height = 3, units = "in")
```

```{r}
Sys.Date. = Sys.Date()
Sys.Date.
session_info. = devtools::session_info()
session_info.
Sys.time() - start_time
```